### ISSUE TITLE
Formly v5 to v6 and Angular v15 to v17 Performance decline

### ISSUE THREAD

**Issue by rothattack** [ngx-formly/ngx-formly#3984]
Hi, we've been using Formly for years now and have not noticed any significant decline
in performance throughout major upgrades with Formly and Angular except recently when
we decided to upgrade Angular to v17, which forced us to upgrade to v6 of Formly.
We now experience UI hanging for seconds, and lots of deep observable firing.
We do have large objects in our formState, and use Handsontable for our datatables wrapped
in formly, but even without tables, there is still a significant pause before render.

Are there any advice or tips i should know or look out for when upgrading from
v5 to v6 of Formly? I read the migration guide and tried some backward compatible configuration
but none have worked so far. Since our app is in Electron, I am unable to use Angular DevTools
and the Performance report show very deep change detection calls.

Any help is appreciated. Thank you.

**Comment by aitboudad** [comment]
Its hard to help without a reproduction, please ensure `checkExpressionOn` is set to `modelChange`:

```patch
FormlyModule.forRoot({
+ extras: {
+   checkExpressionOn: 'modelChange',
+ },
}),
```

if you can share some example that I can check on my side that would be helpful 🙏

**Comment by rothattack** [comment]
Thank you very much for responding! Yes, I have tried configurations mentioned in the migration of v5 to v6.
Would you be open to private messaging for internal information?
BTW, is there a way to disabled the creation of "props" when still using templateOptions? I understand it is deprecated,
but just for the meantime.

**Comment by aitboudad** [comment]
> Would you be open to private messaging for internal information?

My suggest way is to try to make a sample example without sharing any info about your project, but anyway you can reach me using the following email [a.aitboudad@gmail.com](mailto:a.aitboudad@gmail.com)

> BTW, is there a way to disabled the creation of "props" when still using templateOptions?

not clear in v6 `templateOptions` is just a alias to `props` both should work without any side effect.

**Comment by rothattack** [comment]
Thanks. I will try to make a sample and send it to you.

Just for sanity, I am obviously shooting in the dark here,
I build a fields object that has props looking like a normal object:
![image](https://github.com/user-attachments/assets/263ed81f-bac9-45c8-a866-79e5f9e992ef)
But once it gets assigned it converts the properties into getters:
![image](https://github.com/user-attachments/assets/dd9c0400-dbbe-467a-b868-5d2b0008e06b)
Would that affect the performance, triggering a lot of change detection, or no?

Thanks again.

**Comment by aitboudad** [comment]
It could be but not sure, we've a plan to use signal in the upcoming version, try to comment the following line and see if that has an impact in perf https://github.com/ngx-formly/ngx-formly/blob/main/src/core/src/lib/components/formly.field.ts#L287

**Comment by kenisteward** [comment]
@aitboudad 

Internally I'm also seeing some degradation in my wrapper library with even with angular 14 ngx-formly 6.

I'll try to setup a StackBlitz also but it looks like it's mostly and issue with array type fields processing.

**Comment by rothattack** [comment]
> It could be but not sure, we've a plan to use signal in the upcoming version, try to comment the following line and see if that has an impact in perf https://github.com/ngx-formly/ngx-formly/blob/main/src/core/src/lib/components/formly.field.ts#L287

Thanks to your tip i seem to be making progress, would it be wrong to not deeply observe formState? Any consequence of simply doing an observe vs. observeDeep on this line?:
https://github.com/ngx-formly/ngx-formly/blob/84e2150b3ce6acd59c81f5dfa1012cab829376bd/src/core/src/lib/components/formly.field.ts#L290

**Comment by kenisteward** [comment]
> > It could be but not sure, we've a plan to use signal in the upcoming version, try to comment the following line and see if that has an impact in perf https://github.com/ngx-formly/ngx-formly/blob/main/src/core/src/lib/components/formly.field.ts#L287
> 
> Thanks to your tip i seem to be making progress, would it be wrong to not deeply observe formState, our app can have thousands of nested objects in the formState? Any consequence of simply doing an observe vs. observeDeep on this line?:
> https://github.com/ngx-formly/ngx-formly/blob/84e2150b3ce6acd59c81f5dfa1012cab829376bd/src/core/src/lib/components/formly.field.ts#L290

Question:

I'd like to replicate anything you've done to get a quicker perf.  Did you just make changes of what Ait suggested or something else?

**Comment by rothattack** [comment]
> > > It could be but not sure, we've a plan to use signal in the upcoming version, try to comment the following line and see if that has an impact in perf https://github.com/ngx-formly/ngx-formly/blob/main/src/core/src/lib/components/formly.field.ts#L287
> > 
> > 
> > Thanks to your tip i seem to be making progress, would it be wrong to not deeply observe formState, our app can have thousands of nested objects in the formState? Any consequence of simply doing an observe vs. observeDeep on this line?:
> > https://github.com/ngx-formly/ngx-formly/blob/84e2150b3ce6acd59c81f5dfa1012cab829376bd/src/core/src/lib/components/formly.field.ts#L290
> 
> Question:
> 
> I'd like to replicate anything you've done to get a quicker perf. Did you just make changes of what Ait suggested or something else?

Yes, i removed observer for props since we still use templateOptions, simply assign an empty array to:
https://github.com/ngx-formly/ngx-formly/blob/84e2150b3ce6acd59c81f5dfa1012cab829376bd/src/core/src/lib/components/formly.field.ts#L287

And then i do not observeDeep formState, simply observe, since our formState contains a property that contains thousands of object properties in itself:
https://github.com/ngx-formly/ngx-formly/blob/84e2150b3ce6acd59c81f5dfa1012cab829376bd/src/core/src/lib/components/formly.field.ts#L290

This worked for our case, i hope you find what works with yours.

**Comment by kenisteward** [comment]
@aitboudad I've confirmed that changing both of these expressions from observe deep to observe instead brings back performance from before.

I'm in the same boat except on top of having a large form state my clients are trying to render 5000 nodes or more so browser render draw also drags. 

Let me know if there's something I can contribute

**Comment by rothattack** [comment]
Maybe we can add a flag config for shallow observing formState, but it seems we are rare use cases.

i can close this ticket if nobody objects

**Comment by aitboudad** [comment]
If I understood that observing formState is the root reason of the perf decline, I think we should move it to `formly-form` component where it'll be called only once.

**Comment by rothattack** [comment]
I am willing to test that, let me know the details

**Comment by aitboudad** [comment]
This issue has been fixed and released as part of [v6.3.8](https://github.com/ngx-formly/ngx-formly/releases/tag/v6.3.8) release.

Please let us know, in case you are still encountering a similar issue/problem.
Thank you!
